### ISSUE TITLE
Helpful parser error messages

### ISSUE THREAD

**Issue by tzvipm** [tree-sitter/tree-sitter#255]
I can't seem to figure out a way to surface parser errors to the user of my text editor. In particular, I'm working on a VSCode extension using tree-sitter to parse typescript files and surface parsing errors. Tree-sitter is much faster, so far, than the typescript language service, but I can't seem to figure out how to surface a useful error message to the user.

In particular, I'd like something like the `Parsing error: ';' expected.` that eslint gives me. I figure that tree-sitter has some sort of context on the active branches in the parser, so should be able to surface that context int he error object.

If this feature already exists, and I'm just missing something, perhaps it should be included in the documentation site, as I couldn't find anything there.

**Comment by maxbrunsfeld** [comment]
Tree-sitter doesn't currently have this feature, but I agree that it would be a good feature to add.

Right now, Tree-sitter just tries to *repair* the error so that it can give you back a syntax tree that works for the rest of the file.

It's totally possible to use the parse table to generate a message like `Expected tokens: ','  ']'` or something.

**Comment by Razzeee** [comment]
This would be awesome, as I kinda expected that to already work.

**Comment by Razzeee** [comment]
This would be awesome, as I kinda expected that to already work. 

Might also be very helpful for debugging.

**Comment by Razzeee** [comment]
@maxbrunsfeld could you give any pointers on how to engineer this feature? I might take a stab at it.

**Comment by Srekel** [comment]
I would love this. I've just started trying out tree-sitter and the parser error messages have stumped me several times.

Am I correct in understanding that if I actually hooked up the parser and looked at the tree using the API, you get better error information than simply "ERROR"? (rather than just running `--parse` or `test` from the commandline)

**Comment by Arcanemagus** [comment]
> Am I correct in understanding that if I actually hooked up the parser and looked at the tree using the API, you get better error information than simply "ERROR"?

Correct, in fact @aerijo put together a `linter-tree-sitter` package for Atom that takes advantage of this to show Tree-sitter parse errors in the standard linting interface:
https://github.com/Aerijo/linter-tree-sitter/blob/master/lib/linter.js

**Comment by ret2libc** [comment]
Hi, was there any progress on this?

If not, is there any workaround? My grammar accepts things like `<identifier> @ <identifier>` or `<identifier> @a:<identifier>` or `<identifier> @b:<identifier>`, etc. If the user just types `<identifier> @` I would like to show him the possibilities he has (in this case they would be `@`, `@a:`, `@b:`). Is there a way to do this? I think this is the same "problem" as the original issue, if not I will create a new issue.

**Comment by stephe-ada-guru** [comment]
Riccardo Schirone <notifications@github.com> writes:

> My grammar accepts things like
> `<identifier> @ <identifier>` or `<identifier> @a:<identifier>` or
> `<identifier> @b:<identifier>`, etc. If the user just types
> `<identifier> @` I would like to show him the possibilities he has (in
> this case they would be `@`, `@a:`, `@b:`). Is there a way to do this?
> I think this is the same "problem" as the original issue, if not I
> will create a new issue.

For whatever it's worth, the WisiToken
(https://stephe-leake.org/ada/wisitoken.html) parser produces the error
message you are looking for, built from the action list in the current
state of the parse table.

It should be simple to add a similar message to a tree-sitter parser.

-- 
-- Stephe

Written from unceded ancestral homelands of the Huichiun, an Ohlone
people,for which I pay Shuumi Land Tax
(https://sogoreate-landtrust.com/shuumi-land-tax) as an inadequate token
of reparation.

**Comment by hvithrafn** [comment]
@maxbrunsfeld I'm considering trying to implement this functionality but I'm not sure where to start, how difficult it would be, or how useful the end results would be (versus, say, a higher-level custom error reporting implementation for each grammar).

Do you have any thoughts you can share on this?

**Comment by alanz** [comment]
I made something that seems to work, at https://github.com/alanz/tree-sitter/commit/28f835275186e6489df2e2ee50aff10ed947f5fe

**Comment by Razzeee** [comment]
I'm toying with the idea to make tab usage illegal for indention in the elm parser https://github.com/elm-tooling/tree-sitter-elm/issues/97

But I don't think it's a good idea without this feature, as we might not be able to give a decent error message with how tree sitter works right now.

**Comment by theoparis** [comment]
> I made something that seems to work, at [alanz@28f8352](https://github.com/alanz/tree-sitter/commit/28f835275186e6489df2e2ee50aff10ed947f5fe)
> 
> I use it with https://github.com/alanz/haskell-tree-sitter/tree/ts_tree_edit and https://github.com/alanz/semantic/tree/report-diagnostics, in a proof of concept that shows up the diagnostics in a language server.

I have a real world use case for this feature - I could use this in my own language parser, as I want to get a list of errors to send back to the user on failure. Since I'm writing my custom language with Zig tree-sitter seemed to be the best option for writing parser combinators, however I do not want to have to fork alanz's tree-sitter branch just to do so.

**Comment by silvanshade** [comment]
> I could use this in my own language parser, as I want to get a list of errors to send back to the user on failure.

Just to save you some trouble, the branch you reference doesn't really provide information that is suitable for presenting a list of errors like this to the user.

From what I remember, you get a list of `ERROR` (or related) nodes but what you really need is information about *why* those nodes occurred, such as what the parser was expecting versus what it encountered. But tree-sitter still does not provide a way to access this information from the parsing state.

If you want to be able to present this information to users of your parser you will likely need to have a separate syntax validation phase that walks the tree-sitter tree and computes this information. Unfortunately this means you essentially have to write a parser twice, in a sense, because you're just duplicating the same structure you already defined through the tree-sitter rules. But that's about the best you can manage right now.

**Comment by m-kru** [comment]
Is there currently something going on in this matter?

**Comment by andremarianiello** [comment]
I am currently working on a language implementation and I would like to replace my current parser with a tree-sitter parser, but the error messages are going to be tough without contextual information about why the error occurred, so I am looking forward to this feature.

**Comment by kfsone** [comment]
I get the impression that tree-sitter doesn't really try to help you do this at the moment, down to the line/column values being 0-based rather than 1-based.

After looking at the linked linter, it appears that the only real way to do this right now would be to have a grammar that documents all the failure cases and injects error nodes into your tree. Parse the document and run a query to lookup all the error nodes and report them (remember to add 1 to line / column).

The main thing missing is a provide the parser some kind of callback that gets information about errors so that you can use your own mechanism for reporting errors.

**Comment by nicocossiom** [comment]
> After looking at the linked linter, it appears that the only real way to do this right now would be to have a grammar that documents all the failure cases and injects error nodes into 

Yep, can confirm. After sometime searching online I realised the same thing. I am creating a small language with for my compilers class in uni, I need to do type checking as well as showing parsing errors.
For type checking I am traversing the tree, generating symbol tables so i can look up types for identifiers and check expression values. This is tedious but straightforward.

> I get the impression that tree-sitter doesn't really try to help you do this at the moment, down to the line/column values being 0-based rather than 1-based.

It seems like that to a certain point, I think tree-sitter could make it easier to get the parsing errors and things of the sort, but tbh it's quite easy to do, specially in Python: 

```python3
def get_error_nodes(node: Node):
    def traverse_tree_for_errors(node: Node):
        for n in node.children:
            if n.type == "ERROR" or n.is_missing:
                yield n
            if n.has_error:
                # there is an error inside this node let's check inside
                yield from traverse_tree_for_errors(n)

    yield from traverse_tree_for_errors(node)


def print_error_line(er_line: int, padding: str, column_start, column_end, node_error):
    print(f"{padding}{er_line}{padding}{file_lines[er_line-1]}")
    padding_with_line_number = " " * (len(f"{er_line}") + column_start-1)
    cursor_size = max(1, column_end - column_start)
    print(
        f"{padding * 2}{Fore.RED}{padding_with_line_number}{'~' * cursor_size}")

    if node_error.has_error and node_error.is_missing:
        error_message = f"{node_error.sexp()[1:-1]}"
    else:
        unexpected_tokens = "".join(n.text.decode('utf-8')
                                    for n in node_error.children)
        error_message = f"Unexpected token(s): {unexpected_tokens}"
    print(
        f"{padding * 2}{Fore.RED}{padding_with_line_number}{error_message}:")


def print_error(root_node: Node, error_type: str = "SYNTAX_ERROR"):
    padding = " " * 5
    for node_error in get_error_nodes(root_node):
        er_line = node_error.start_point[0]+1
        column_start = node_error.start_point[1] + 1
        column_end = node_error.end_point[1] + 1
        print(
            f"{Fore.RED}{error_type}{Fore.RESET}:  {node_error.sexp()[1:-1]}")
        print(
            f"{padding}in file: '{file}:{er_line}:{column_start}:{column_end}', line: {er_line}", end=", ")
        print(
            f"from column {column_start} to {column_end}\n")
        print(f"{padding}{file_name}")
        if "--show-file" in sys.argv:
            print_file_with_errors(er_line, padding,
                                   column_start, column_end, node_error)
        else:
            print_error_line(er_line, padding, column_start,
                             column_end, node_error)
```

an example:
```jspdl
let string somestring /* missing ; */

do{
    input( a + b ); /* input only accepts an identifier as an argument */
}while(); /* Missing expression */
```
```txt
 python main.py jspdl_language/test2.jspdl                                               (tree-sitter) 
└── program
    ├── let_statement
    │   ├── let
    │   ├── type
    │   │   └── string
    │   ├── identifier
    │   └── ;
    └── do_while_statement
        ├── do
        ├── {
        ├── input_statement
        │   ├── input
        │   ├── (
        │   ├── identifier
        │   ├── ERROR
        │   │   ├── +
        │   │   └── identifier
        │   ├── )
        │   └── ;
        ├── }
        ├── while
        ├── (
        └── value
            └── expression_value
                └── literal_string
        ├── )
        └── ;
SYNTAX_ERROR:  MISSING ";"
     in file: 'jspdl_language/test2.jspdl:1:22:22', line: 1, from column 22 to 22

     /Users/pepperonico/Developer/upm/TdL/jspdl_language/test2.jspdl
     1     let string somestring
                                ~
                                MISSING ";":
SYNTAX_ERROR:  ERROR (identifier)
     in file: 'jspdl_language/test2.jspdl:4:14:17', line: 4, from column 14 to 17

     /Users/pepperonico/Developer/upm/TdL/jspdl_language/test2.jspdl
     4         input( a + b);
                        ~~~
                        Unexpected token(s): +b:
SYNTAX_ERROR:  MISSING "literal_string"
     in file: 'jspdl_language/test2.jspdl:5:8:8', line: 5, from column 8 to 8

     /Users/pepperonico/Developer/upm/TdL/jspdl_language/test2.jspdl
     5     }while();
                  ~
                  MISSING "literal_string":
```
Its weird how tree-sitter generates parsing errors. For example in the last one as can bee seen in the tree it assumed that inside the while() there should have been a string bc it's the first rule under the value rule which in turn is the first rule under the expression rule(see attached grammar.js). Idk this seems counterintuitive. I would assume in this cases that it would say that the missing rule is an expression (as seen in the ast it is what it should be), but it went on to assume that if there was to be something there one of the rules that might fit is a string.

I understand the premise of tree-sitter try to parse a file and be able to recover from missing rules or unfitting tokens. But imo it assumes too mu
ch, maybe I don't fully understand how tree-sitter works but i think that it could just assume that there's an expression and keep going, it assumed up to 2 more rules which was just completely unnecessary. Again I am a newbie and maybe my grammar is not well defined.
[grammar.js.zip](https://github.com/tree-sitter/tree-sitter/files/11584607/grammar.js.zip)

**Comment by ElectronicRU** [comment]
One workable approach could be storing a parser-state(s?) in MISSING nodes, and augmenting table generator with info such as named nonterminals + terminals that are valid to be shifted into that state(s). Together with inspecting the node's parent, I think that would be workable information to produce good errors.

New to the project, however, so hard for me to determine how feasible that idea is. Should I take a crack at an implementation, or would I waste my time?

**Comment by ahlinc** [comment]
API for this is already on the master branch.

**Comment by pherrymason** [comment]
@ahlinc , I've been reading the docs and cluld not find any api related to errors. Could you point me where this api is? A link to sources is ok

**Comment by sogaiu** [comment]
@pherrymason [FYI](https://github.com/tree-sitter/tree-sitter/issues/2797#issuecomment-1912930142)

**Comment by Termina1** [comment]
But is the API in the master?

**Comment by ObserverOfTime** [comment]
Yes, see https://github.com/tree-sitter/tree-sitter/pull/2324#issuecomment-1637468158. We do need an example in the docs though.
